crates/sui-graphql-rpc/src/types/transaction_block/tx_lookups.rs

```diff
@@ -0,0 +1,468 @@
+#[macro_export]
+macro_rules! min_option {
+    ($($x:expr),+ $(,)?) => {{
+        [$($x),*].iter()
+            .filter_map(|&x| x)
+            .min()
+    }};
+}
+
+/// Constructs a `RawQuery` as a join over all relevant side tables, filtered on their own filter
+/// condition, plus optionally a sender, plus optionally tx/cp bounds.
+pub(crate) fn subqueries(filter: &TransactionBlockFilter, tx_bounds: TxBounds) -> Option<RawQuery> {
+    let sender = filter.sign_address;
+
+    let mut subqueries = vec![];
+
+    if let Some(f) = &filter.function {
+        subqueries.push(match f {
+            FqNameFilter::ByModule(filter) => match filter {
+                ModuleFilter::ByPackage(p) => (select_pkg(p, sender, tx_bounds), "tx_calls_pkg"),
+                ModuleFilter::ByModule(p, m) => {
+                    (select_mod(p, m.clone(), sender, tx_bounds), "tx_calls_mod")
+                }
+            },
+            FqNameFilter::ByFqName(p, m, n) => (
+                select_fun(p, m.clone(), n.clone(), sender, tx_bounds),
+                "tx_calls_fun",
+            ),
+        });
+    }
+    if let Some(kind) = &filter.kind {
+        subqueries.push((select_kind(*kind, tx_bounds), "tx_kinds"));
+    }
+    if let Some(recv) = &filter.recv_address {
+        subqueries.push((select_recipient(recv, sender, tx_bounds), "tx_recipients"));
+    }
+    if let Some(input) = &filter.input_object {
+        subqueries.push((select_input(input, sender, tx_bounds), "tx_input_objects"));
+    }
+    if let Some(changed) = &filter.changed_object {
+        subqueries.push((
+            select_changed(changed, sender, tx_bounds),
+            "tx_changed_objects",
+        ));
+    }
+    if let Some(sender) = &sender {
+        if filter.complex_filters() == 0 || filter.kind.is_some() {
+            subqueries.push((select_sender(sender, tx_bounds), "tx_senders"));
+        }
+    }
+    if let Some(txs) = &filter.transaction_ids {
+        subqueries.push((select_ids(txs, tx_bounds), "tx_digests"));
+    }
+
+    if subqueries.is_empty() {
+        return None;
+    }
+
+    let mut subquery = subqueries.pop().unwrap().0;
+
+    if !subqueries.is_empty() {
+        subquery = query!("SELECT tx_sequence_number FROM ({}) AS initial", subquery);
+        while let Some((subselect, alias)) = subqueries.pop() {
+            subquery = inner_join!(subquery, alias => subselect, using: ["tx_sequence_number"]);
+        }
+    }
+
+    Some(subquery)
+}
+
+pub(crate) fn select_tx(sender: Option<SuiAddress>, bound: TxBounds, from: &str) -> RawQuery {
+    let mut query = query!(format!("SELECT tx_sequence_number FROM {}", from));
+
+    if let Some(sender) = sender {
+        query = filter!(query, format!("sender = {}", bytea_literal(&sender)));
+    }
+
+    filter!(
+        query,
+        format!(
+            "tx_sequence_number >= {} AND tx_sequence_number <= {}",
+            bound.lo, bound.hi
+        )
+    )
```

Comment:
nits:

- Can't we use SQL binds for `bound.lo` and `bound.hi`?
- We can structure this to be slightly simpler by applying the tx sequence number constraint first.
- Keep chains of inequalities monotonic, to make them easier to follow.

```suggestion
    let mut query = filter!(
        query!(format!("SELECT tx_sequence_number FROM {}", from)),
        "{} <= tx_sequence_number AND tx_sequence_number <= {}",
        bounds.lo,
        bounds.hi,
    );

    if let Some(sender) = sender {
        query = filter!(query, format!("sender = {}", bytea_literal(&sender)));
    }

    query
```

Reply:
(I'm vaguely remembering that we can't use binds, because binds are just for strings)

Reply:
`fn filter` is implemented on `T: Display` so I think it might/ should work for `bounds.lo` and `bounds.hi`? Otherwise we can still adopt your suggestion, but just with `format!("{} <= tx...", bounds.lo, bounds.hi)`